database/src/main/java/com/firebase/ui/database/FirebaseArrayOfObjects.java

```diff
@@ -0,0 +1,237 @@
+package com.firebase.ui.database;
+
+import android.util.Pair;
+
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+/**
+ * Acts as a bridge between a list of {@link DataSnapshot}s and a list of objects of type E.
+ *
+ * @param <E> the model representation of a {@link DataSnapshot}
+ */
+public class FirebaseArrayOfObjects<E> extends ImmutableList<E> {
+    protected List<DataSnapshot> mSnapshots;
+    protected Class<E> mEClass;
+    protected SnapshotParser<E> mParser;
+
+    /**
+     * @param snapshots  a list of {@link DataSnapshot}s to be converted to a model type
+     * @param modelClass the model representation of a {@link DataSnapshot}
+     */
+    public FirebaseArrayOfObjects(List<DataSnapshot> snapshots, Class<E> modelClass) {
```

Comment:
Should this constructor be private?  Seems strange to have a public constructor and `newInstance`